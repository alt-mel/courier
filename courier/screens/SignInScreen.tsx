import React, { useState, useEffect } from 'react';
import { Pressable, Alert } from 'react-native';
import { View, Text, TextInput } from '../components/Themed';

import { useNavigation } from '@react-navigation/native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { useMutation, gql } from '@apollo/client';

const SIGN_IN_MUTATION = gql`
  mutation signIn($email: String!, $password: String!) {
    signIn(input: { email: $email, password: $password }) {
      token
      user {
        id
        name
        email
      }
    }
  }
`;

const SignInScreen = () => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  const navigation = useNavigation();

  const [signIn, { data, error, loading }] = useMutation(SIGN_IN_MUTATION);

  useEffect(() => {
    if (error) {
      Alert.alert('Invalid credentials, try again');
    }
  }, [error]);

  if (data) {
    AsyncStorage.setItem('token', data.signIn.token).then(() => {
      navigation.navigate('Home');
    });
  }

  const onSubmit = () => {
    signIn({ variables: { email, password } });
  };

  return (
    <View style={{ padding: 20 }}>
      <TextInput
        placeholder="email@email.com"
        value={email}
        onChangeText={setEmail}
        style={{
          fontSize: 18,
          width: '100%',
          marginVertical: 25,
        }}
      />

      <TextInput
        placeholder="password"
        value={password}
        onChangeText={setPassword}
        secureTextEntry
        style={{
          fontSize: 18,
          width: '100%',
          marginVertical: 25,
        }}
      />
      <Pressable
        onPress={onSubmit}
        disabled={loading}
        style={{
          backgroundColor: '#ed706E',
          height: 50,
          borderRadius: 5,
          alignItems: 'center',
          justifyContent: 'center',
          marginTop: 30,
        }}
      >
        <Text
          style={{
            color: 'white',
            fontSize: 18,
            fontWeight: 'bold',
          }}
        >
          Sign In
        </Text>
      </Pressable>

      <Pressable
        onPress={() => {
          console.warn('nbavigate');
          navigation.navigate('SignUp');
        }}
        style={{
          height: 50,
          borderRadius: 5,
          alignItems: 'center',
          justifyContent: 'center',
          marginTop: 30,
        }}
      >
        <Text
          style={{
            color: '#ed706E',
            fontSize: 18,
            fontWeight: 'bold',
          }}
        >
          New here? Sign up
        </Text>
      </Pressable>
    </View>
  );
};

export default SignInScreen;
