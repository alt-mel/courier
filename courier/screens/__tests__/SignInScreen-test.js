import * as React from 'react';
import renderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';
import { gql } from '@apollo/client';
import {render, fireEvent, act } from '@testing-library/react-native';
import SignInScreen from '../SignInScreen';

const SIGN_IN_MUTATION = gql`
  mutation signIn($email: String!, $password: String!) {
    signIn(input: { email: $email, password: $password }) {
      token
      user {
        id
        name
        email
      }
    }
  }
`;

const mocks = [ {
    request: {
      query: SIGN_IN_MUTATION,
      variables: {
        email: 'email',
        password: 'password'
      },
    },
    result: {
      data: {
        signIn: {
          token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjYxMzYzOTc0YjZmMjZjNmNkNzg4MTE4MyIsImlhdCI6MTYzMTEyNTg4MSwiZXhwIjoxNjMzNzE3ODgxfQ.QlAWUoJ6z-40MS2GtDsbc9btN98Lxb_Nd4Z4I3DPGEA',
          user: {
            id: '61363974b6f26c6cd7881183',
            name: 'John Briggs',
            email: 'admin@john.com'
          }
        }
      }
    },
  },];

it('renders the Sign In Screen', () => {
  const tree = renderer.create( <MockedProvider><SignInScreen/></MockedProvider>).toJSON();
  expect(tree).toMatchSnapshot();
});


it('shows invalid credentials message', async () => {
    const {getByTestId, queryAllByText } = render(<MockedProvider mocks={mocks} addTypename={false}><SignInScreen/></MockedProvider> );
  
     act(() =>
        fireEvent.changeText(getByTestId('SignIn.Email'), 'skadjsk'),
        fireEvent.changeText(getByTestId('SignIn.Password'), 'skadjsk'),    
        fireEvent.press(getByTestId('SignIn.Button'))
    );
    
    expect(queryAllByText('Invalid credentials, try again')).toBe(1);  
  });