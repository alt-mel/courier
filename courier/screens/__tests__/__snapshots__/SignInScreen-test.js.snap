// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Sign In Screen 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "padding": 20,
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="email@email.com"
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "color": "#000",
        },
        Object {
          "fontSize": 18,
          "marginVertical": 25,
          "width": "100%",
        },
      ]
    }
    testID="SignIn.Email"
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="password"
    rejectResponderTermination={true}
    secureTextEntry={true}
    style={
      Array [
        Object {
          "color": "#000",
        },
        Object {
          "fontSize": 18,
          "marginVertical": 25,
          "width": "100%",
        },
      ]
    }
    testID="SignIn.Password"
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ed706E",
        "borderRadius": 5,
        "height": 50,
        "justifyContent": "center",
        "marginTop": 30,
      }
    }
    testID="SignIn.Button"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
          },
          Object {
            "color": "white",
            "fontSize": 18,
            "fontWeight": "bold",
          },
        ]
      }
    >
      Sign In
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 5,
        "height": 50,
        "justifyContent": "center",
        "marginTop": 30,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
          },
          Object {
            "color": "#ed706E",
            "fontSize": 18,
            "fontWeight": "bold",
          },
        ]
      }
    >
      New here? Sign up
    </Text>
  </View>
</View>
`;
